examples/vae/train_vae.py

```diff
@@ -117,33 +118,57 @@ def main():
     # Run the training
     trainer.run()
 
-    # Visualize the results
-    def save_images(x, filename):
-        import matplotlib.pyplot as plt
-        fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
-        for ai, xi in zip(ax.flatten(), x):
-            ai.imshow(xi.reshape(28, 28))
-        fig.savefig(filename)
+    # Save images for demonstration
+    save_images(device, encoder, decoder, train, test, prior, args.out)
 
-    avg_elbo_loss.to_cpu()
-    train_ind = [1, 3, 5, 10, 2, 0, 13, 15, 17]
-    x = chainer.Variable(np.asarray(train[train_ind]))
-    with chainer.using_config('train', False), chainer.no_backprop_mode():
-        x1 = decoder(encoder(x).mean, inference=True).mean
-    save_images(x.array, os.path.join(args.out, 'train'))
-    save_images(x1.array, os.path.join(args.out, 'train_reconstructed'))
 
-    test_ind = [3, 2, 1, 18, 4, 8, 11, 17, 61]
-    x = chainer.Variable(np.asarray(test[test_ind]))
-    with chainer.using_config('train', False), chainer.no_backprop_mode():
-        x1 = decoder(encoder(x).mean, inference=True).mean
-    save_images(x.array, os.path.join(args.out, 'test'))
-    save_images(x1.array, os.path.join(args.out, 'test_reconstructed'))
+# Saves 3x3 tiled image
+def save3x3(x, filename):
+    numpy_device = chainer.get_device('@numpy')
+    fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
+    for ai, xi in zip(ax.flatten(), x):
+        im = xi.reshape(28, 28)
+        im = numpy_device.send(im)
+        ai.imshow(im)
+    fig.savefig(filename)
+
 
-    # draw images from randomly sampled z
+# Saves reconstruction images using:
+# - training image samples
+# - test image samples
+# - randomly sampled values of z
+def save_images(device, encoder, decoder, train, test, prior, out_dir):
+
+    # Training samples
+    train_ind = [1, 3, 5, 10, 2, 0, 13, 15, 17]
+    x_numpy = train[train_ind]
+    x = device.send(x_numpy)
+    with chainer.using_config('train', False):
+        with chainer.no_backprop_mode():
```

Comment:
I thought it's easier to understand.
It's difficult to notice there are two context managers if they're written in a single line.
It may be fine to write them in a single `with` statement, but I believe they should be in separate lines.
By the way, I don't understand what `train=False` is targeting.
We need comments to both context managers.
(Note that intended audience is those who have never read code written using Chainer.)

Reply:
> By the way, I don't understand what `train=False` is targeting.

I found @mitmul proposed adding `train=False` (https://github.com/chainer/chainer/pull/4009#discussion_r155265234).

Reply:
Before that, the `train` flag (for educational purposes) was removed once (https://github.com/chainer/chainer/pull/2186/files#diff-ba800ec076cde92acec7087afdf866e5).

Reply:
I see.
Honestly I don't think we should apply `train` flag for such reason.
If you modify the model, you must adjust the configuration as required by the new model. That's it.

But it seems to be out of scope of this PR.
As for the coding style of multiple context managers, what's your opinion?

Reply:
Both styles are readable to me.  I'll follow your decision.